### PR TITLE
Width and Height

### PR DESCRIPTION
**Pull request by ggovardh**
## Related Issue
Please use one of the well-known [github fixes keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to reference
the issues fixed with this PR (eg Fixes #<github issue number>). If an issue doesn't yet exist please create one if reasonable to aid 
in issue tracking.

**NOTE**: For multiple issues resolved by this PR use the corresponding keywords **every time** in a comma-delimited list per the reference
page above.

## Description
Create a [descriptive title and provide all of the relevant details in the description](https://chris.beams.io/posts/git-commit/). 
This information lives with the code and will help future readers (***including yourself***) and will serve as documentation.

At the very least please describe the issue you are addressing, and what your fix entails. 

## Sample Card
image
image
image
image
image
image

## How Verified
How you verified the fix, including one or all of the following: 
1. New unit tests that were added if any. If none were added please add a quick line explaining why not.
2. Existing relevant unit/regression tests that you ran
3. Manual scenario verification if any; ***Do include .gif's or screenshots of the testing you performed here if you think that it 
will aid in code reviews or corresponding fixes on other platforms for eg.***

### CODE REVIEW

File: source/shared/cpp/ObjectModel/CollectionTypeElement.cpp [hunk 1]
```diff
@@ -46,7 +46,7 @@ void CollectionTypeElement::SetPadding(const bool value)
 void CollectionTypeElement::ConfigPadding(const ParseContext& context)
 {
     // We set padding when child's style is set explicitly (not None) and is different than the parental style or if the style is emphasis or if background image is not present
-    const bool padding = ( GetBackgroundImage() != NULL || ((GetStyle() != ContainerStyle::None) && (context.GetParentalContainerStyle()!= GetStyle())) || GetStyle() == ContainerStyle::Emphasis);
+    const bool padding = ((GetBackgroundImage() == nullptr && GetStyle() != ContainerStyle::None && (context.GetParentalContainerStyle()!= GetStyle())) || GetStyle() == ContainerStyle::Emphasis);
```
Comment (rj100189) [inline source/shared/cpp/ObjectModel/CollectionTypeElement.cpp:49]: what is this change intended for?

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 2]
```diff
@@ -106,9 +106,22 @@ namespace RendererQml
         AddContainerElements(bodyLayout, card->GetBody(), context);
         AddActions(bodyLayout, card->GetActions(), context);
 
+		bodyLayout->Property("onHeightChanged", Formatter() << "{generateStretchHeight(bodyLayout.children," << card->GetMinHeight() << ");}");
+
+		bodyLayout->Property("onWidthChanged", Formatter() << "{generateStretchHeight(bodyLayout.children," << card->GetMinHeight() << ")}");
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:111]: `generateStretchHeight` -> is this name correct here? you are check for `onWidthChanged`.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 3]
```diff
@@ -126,6 +139,11 @@ namespace RendererQml
 				}
 
 				//TODO: Add collection element
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:141]: Is this TODO still valid?

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 4]
```diff
@@ -149,6 +167,8 @@ namespace RendererQml
                 uiButtonStrip->Property("width", "parent.width");
                 uiButtonStrip->Property("spacing", std::to_string(actionsConfig.buttonSpacing));
 
+				
+
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:171]: remove this extra line

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 5]
```diff
@@ -106,9 +106,22 @@ namespace RendererQml
...
+		bodyLayout->Property("onWidthChanged", Formatter() << "{generateStretchHeight(bodyLayout.children," << card->GetMinHeight() << ")}");
+
+		bodyLayout->AddFunctions(AdaptiveCardQmlRenderer::getStretchHeight());
+
+		bodyLayout->AddFunctions(AdaptiveCardQmlRenderer::getStretchWidth());
+
+		bodyLayout->AddFunctions(AdaptiveCardQmlRenderer::getMinWidth());
+
+		bodyLayout->AddFunctions(AdaptiveCardQmlRenderer::getMinWidthActionSet());
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:119]: Should these function be inside `bodyLayout`? Can we have it in root itself... and then we can use `context->getCardRootId` to get the id wherever we need to call the function.

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 6]
```diff
@@ -1408,6 +1441,23 @@ namespace RendererQml
...
+			tempMargin = context->GetConfig()->GetSpacing().paddingSpacing;
+		}
+
+		uiColumn->Property("onHeightChanged", Formatter() << "{bodyLayout.generateStretchHeight(children, " << id << ".minHeight - " << 2 * tempMargin << " )}");
+
+		uiColumn->Property("onWidthChanged", Formatter() << "{bodyLayout.generateStretchHeight(children, " << id << ".minHeight - " << 2 * tempMargin << " )}");
+
+		uiContainer->Property("onMinHeightChanged", Formatter() << "{bodyLayout.generateStretchHeight( column_" << id << ".children, minHeight - " << 2 * tempMargin << " )}");
+
+		uiContainer->Property("property int minWidth", Formatter() << "bodyLayout.getMinWidth( column_" << cardElement->GetId() << ".children) + " << 2 * tempMargin);
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:1457]: so, once we make the above change, here we can call `context->getCardRootId().generate......` instead of hardcoding as `bodyLayout`

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 7]
```diff
@@ -2267,5 +2302,195 @@ namespace RendererQml
...
         return function.str();
     }
+
+	const std::string RendererQml::AdaptiveCardQmlRenderer::getStretchHeight()
+	{
+		return R"(
+		function generateStretchHeight(childrens,minHeight)
+        {
+            var n = childrens.length
+            var implicitHei = 0;
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:2312]:
```suggestion
            var implicitHt = 0;
```

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp [hunk 8]
```diff
@@ -2267,5 +2302,195 @@ namespace RendererQml
 
         return function.str();
     }
+
+	const std::string RendererQml::AdaptiveCardQmlRenderer::getStretchHeight()
```
Comment (rj100189) [inline source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:2306]: Is it possible to format this function a bit for easy readability?

**Review by rj100189 — CHANGES_REQUESTED** [review]

File: source/shared/cpp/ObjectModel/CollectionTypeElement.cpp (hunk 1 above)
Comment (ggovardh) [reply source/shared/cpp/ObjectModel/CollectionTypeElement.cpp:49]: The condition for apply padding was not fixed correctly the last time and was creating problem when background image was present. So the conditions are checked correctly this time

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 2 above)
Comment (ggovardh) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:111]: Yes, Had to apply generateStretchHeight() for both onHeightChanged and onWidthChanged. And since the bodyLayout is not column set generateStretchWidth() is not required

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 5 above)
Comment (ggovardh) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:119]: Yes we can have the functions inside the root layout

File: source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp (hunk 8 above)
Comment (ggovardh) [reply source/qml/Library/RendererQml/AdaptiveCardQmlRenderer.cpp:2306]: I will try to format these function

File: source/shared/cpp/ObjectModel/CollectionTypeElement.cpp (hunk 1 above)
Comment (rj100189) [reply source/shared/cpp/ObjectModel/CollectionTypeElement.cpp:49]: This change is applied in the shared cpp.... I think we should discuss the same with the Mac and MS team as well.

Comment (ggovardh) [reply source/shared/cpp/ObjectModel/CollectionTypeElement.cpp:49]: I will cross check once again. The last time also the conditions were modified by me based on Apply Padding from HTML Renderer

File: source/shared/cpp/ObjectModel/CollectionTypeElement.cpp [hunk 9]
```diff
@@ -46,7 +46,7 @@ void CollectionTypeElement::SetPadding(const bool value)
 void CollectionTypeElement::ConfigPadding(const ParseContext& context)
 {
     // We set padding when child's style is set explicitly (not None) and is different than the parental style or if the style is emphasis or if background image is not present
-    const bool padding = ( GetBackgroundImage() != NULL || ((GetStyle() != ContainerStyle::None) && (context.GetParentalContainerStyle()!= GetStyle())) || GetStyle() == ContainerStyle::Emphasis);
+    const bool padding = (GetBackgroundImage() != NULL || ((GetStyle() != ContainerStyle::None) && (context.GetParentalContainerStyle() != GetStyle())) || GetStyle() == ContainerStyle::Emphasis);
```
Comment (rj100189) [inline source/shared/cpp/ObjectModel/CollectionTypeElement.cpp:49]:
```suggestion
    const bool padding = ( GetBackgroundImage() != NULL || ((GetStyle() != ContainerStyle::None) && (context.GetParentalContainerStyle()!= GetStyle())) || GetStyle() == ContainerStyle::Emphasis);
```

**Review by rj100189 — APPROVED** [review]
